lib/Controller/PasswordController.php

```diff
@@ -0,0 +1,167 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\PasswordPolicy\Controller;
+
+use OCP\AppFramework\Controller;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\JSONResponse;
+use OCP\AppFramework\Http\RedirectResponse;
+use OCP\AppFramework\Http\TemplateResponse;
+use OCP\Authentication\IAccountModuleController;
+use OCP\IConfig;
+use OCP\IL10N;
+use OCP\IRequest;
+use OCP\ISession;
+use OCP\IURLGenerator;
+use OCP\IUserManager;
+use OCP\IUserSession;
+
+class PasswordController extends Controller implements IAccountModuleController {
+
+	/** @var IUserSession */
+	protected $userSession;
+	/** @var IUserManager */
+	protected $userManager;
+	/** @var IConfig */
+	protected $config;
+	/** @var ISession */
+	protected $session;
+	/** @var IURLGenerator */
+	protected $urlGenerator;
+	/** @var IL10N */
+	protected $l10n;
+
+	/**
+	 * PasswordController constructor.
+	 *
+	 * @param $appName
+	 * @param IRequest $request
+	 * @param IUserSession $userSession
+	 * @param IUserManager $userManager
+	 * @param IConfig $config
+	 * @param ISession $session
+	 * @param IURLGenerator $urlGenerator
+	 * @param IL10N $l10n
+	 */
+	public function __construct(
+		$appName,
+		IRequest $request,
+		IUserSession $userSession,
+		IUserManager $userManager,
+		IConfig $config,
+		ISession $session,
+		IURLGenerator $urlGenerator,
+		IL10N $l10n
+	) {
+		parent::__construct($appName, $request);
+		$this->userSession = $userSession;
+		$this->userManager = $userManager;
+		$this->config = $config;
+		$this->session = $session;
+		$this->urlGenerator = $urlGenerator;
+		$this->l10n = $l10n;
+	}
+
+	/**
+	 * @param $redirect_url
+	 * @param string $error (optional)
+	 * @return TemplateResponse
+	 */
+	private function createPasswordTemplateResponse($redirect_url, $error = null) {
+		$params = [
+			'redirect_url' => $redirect_url,
+			'webroot' => \OC::$WEBROOT
+		];
+		if ($error) {
+			$params['error'] = $error;
+		}
+		return new TemplateResponse(
+			$this->appName,
+			'password',
+			$params,
+			'guest'
+		);
+	}
+	/**
+	 * Shows a change password template
+	 *
+	 * @param string $redirect_url
+	 * @return TemplateResponse|RedirectResponse|JSONResponse
+	 *
+	 * @throws \Exception
+	 */
+	public function show($redirect_url) {
+		// gracefully handle redirected ocs api requests
+		if ($this->request->getHeader('OCS-APIREQUEST') === 'true') {
+			return new JSONResponse([], Http::STATUS_NO_CONTENT);
+		}
+		return $this->createPasswordTemplateResponse($redirect_url);
+	}
+
+	/**
+	 * Shows a change password template
+	 *
+	 * @param string $current_password
+	 * @param string $new_password
+	 * @param string $retyped_password
+	 * @param string $redirect_url
+	 * @return TemplateResponse|RedirectResponse
+	 *
+	 * @UseSession
+	 * @throws \Exception
+	 */
+	public function update($current_password, $new_password, $retyped_password, $redirect_url) {
```

Comment:
handle redirect url being null

Reply:
cannot be null because the account module always adds a redirect url

Reply:
hehe - again - I mean on the show method ;-)

Reply:
and show is called by the redirect request, which will contain a redirect url. if it doesn't someone is fiddling with the requests. he will see the force password change template ... but will not be redirected anywhere ... ok from my pov

Reply:
I can call show be entering the url in the browser - after changing the password I get a white page because the redirect holds no value.

Reply:
you are not supposed to enter that url in the browser

Reply:
the acceptance test will call the url directly ;-)
just add the default redirect url - we have this value
without setting the redirect url the redirect response is not valid